### PR TITLE
[WIP] vmware: try to find exact file when multiple files are returned

### PR DESCRIPTION
**Pull request by rhtyd**
On searching a file in the datastore when multiple files are found
by the datastore MO, try and loop through the result and find the file
exactly matching the requested file name.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by rhtyd** [comment]
cc @PaulAngus @andrijapanicsb @DaanHoogland 
This is a work-in-progress PR, the idea is to ensure that when multiple results are returned while searching a datastore, it tries to further loop and find exact-ness in the file being searched. I'm not sure if that breaks any feature for older styled VMs/volumes/snapshots etc.

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/DatastoreMO.java [hunk 1]
```diff
@@ -416,22 +414,39 @@ public String searchFileInSubFolders(String fileName, boolean caseInsensitive, S
...
             if (info != null && info.size() > 0) {
                 for (FileInfo fi : info) {
-                    absoluteFileName = parentFolderPath = result.getFolderPath();
-                    s_logger.info("Found file " + fileName + " in datastore at " + absoluteFileName);
-                    if (parentFolderPath.endsWith("]"))
-                        absoluteFileName += " ";
-                    absoluteFileName += fi.getPath();
-                    if(isValidCloudStackFolderPath(parentFolderPath, searchExcludedFolders)) {
+                    String possibleFilename = null;
+                    possibleFilename = parentFolderPath = result.getFolderPath();
```
Comment (DaanHoogland) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/DatastoreMO.java]: should this double assignment really be here? it does not make the code clearer.

Comment (rhtyd) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/DatastoreMO.java]: it's copied/refactored code (nothing new I wrote mostly), see old code line 423:
```
                    absoluteFileName = parentFolderPath = result.getFolderPath();
```

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
Can you pls give/link some example (I recall related inhouse issues...) on how to test this (i.e. example of the problem)?

**Comment by rhtyd** [comment]
@andrijapanicsb I think it has to do with a VM with many number of data disks, and we perform vmotion/storage migration that changes the volume names (this was from a support issue).

**Comment by rhtyd** [comment]
@andrijapanicsb  I think the issue was of deleting a volume or a snapshot or a failed snapshot? or something (sorry don't recall) where the code is written in such a way that it might find the first file matching the string instead of an exact match and could potentially remove a useful vmdk/file.
@borisstoyanov @PaulAngus @DaanHoogland @nvazquez do you remember anything?

**Comment by DaanHoogland** [comment]
@andrijapanicsb @rhtyd This was noticed when Vmware would rename disk to ROOT-000001.vmdk when the hundred thousand mark was reached. The disks would no longer have a unique name and the wrong disk would be removed.

**Comment by rhtyd** [comment]
Yes, I remember faintly now thanks @DaanHoogland - does that give you enough to test it @andrijapanicsb @borisstoyanov @PaulAngus ?

**Comment by andrijapanicsb** [comment]
not really - I can experiment with trying to reproduce first, but that's gonna take time.

**Comment by rhtyd** [comment]
@andrijapanicsb @DaanHoogland @PaulAngus @borisstoyanov - I'm not sure how we should test this, should we push to 4.15 instead if we're not sure?

**Comment by DaanHoogland** [comment]
@rhtyd, this can be tested by setting in value = '999999' the record sequence with name == 'vm_instance_seq' and than create a couple of VMs with disks. It is quite elaborate (manual setup), so if we want to spend the time fine, otherwise push to 4.15.

**Comment by rhtyd** [comment]
Okay I've pushed it to 4.15 for now, given we've a lot of PRs/issues to deal with. (maybe we can revisit this for a 4.14.1/* in future)

**Comment by rhtyd** [comment]
@harikrishna-patnala with your VMware PR, is this still valid - that is, is there code to fuzzy search vmdk files (instead of exact search). In which case is something like this PR still valid?

**Comment by rhtyd** [comment]
Closing - this may not be necessary with recent move to object/storage manager where low-level operations are not needed/performed.

**Pull request closed without merging**